src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -585,28 +598,45 @@ public Object compactObjects(DynamicObject array,
     @CoreMethod(names = "compact!", raiseIfFrozenSelf = true)
     public abstract static class CompactBangNode extends ArrayCoreMethodNode {
 
-        @Specialization(guards = { "strategy.matches(array)", "!strategy.accepts(nil())" }, limit = "ARRAY_STRATEGIES")
+        @Specialization(guards = { "strategy.matches(array)", "strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
         public DynamicObject compactNotObjects(DynamicObject array,
                 @Cached("of(array)") ArrayStrategy strategy) {
             return nil();
         }
 
-        @Specialization(guards = { "strategy.matches(array)", "strategy.accepts(nil())" }, limit = "ARRAY_STRATEGIES")
+        @Specialization(guards = { "strategy.isStorageMutable()", "strategy.matches(array)", "!strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
         public Object compactObjects(DynamicObject array,
                 @Cached("of(array)") ArrayStrategy strategy) {
+            return compactElements(array, strategy, strategy);
+        }
+
+        @Specialization(guards = { "!strategy.isStorageMutable()", "strategy.matches(array)", "!strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
+        public Object compactObjectsNonMutable(DynamicObject array,
+                @Cached("of(array)") ArrayStrategy strategy,
+                @Cached("strategy.generalizeForMutation()") ArrayStrategy mutableStrategy) {
```

Comment:
Can we just call generalizeForMutation() all the time and have return the same strategy if `isStorageMutable()`?